Standardize the MMR music file format to align with the OOTR music file format

crinuleiroz opened the pull request:
### List of Changes
The majority of this update is updating the MMR music file standards to align with the OOTR music file standard. However, there's a README update as well as a change that lets the music folder search for music files as deep as it needs to tacked onto it.

#### Major Changes
> [!NOTE]
> Changes are considered major if they are affected by the music format update.

- **Updated the `.mmrs` file format to align with the `.ootrs` file format**
    - Filenames are no longer used to link files
    - Filenames no longer contain any metadata
    - Metadata (Display name, instrument set, categories, and custom audio samples) is now stored in a singular metadata (`.meta`) file.
    - Sequence files now use *only* the `.seq` file extension
    - Having multiple sequence files and bank files is no longer supported
- **Legacy music file formats are no longer supported**
    - `.zseq` files will be ignored
    - `.mmrs` files using the previous format will be ignored
- **Added `DisplayName` field to the `SequenceInfo` class**
    - The `DisplayName` is now used for the in-game display name. This allows plando, spoiler log, and songtest & songforce to continue using `Name` (filename) instead of relying on the cosmetic name in the `.meta` file.
- **Added a music conversion utility script to the randomizer**
    - Backs up the music folder into a `music.old` zip archive
    - Converts any and all of the previously supported MMR music file formats to the new music file format
    - Extra files are retained in both the music folder and music files
    - The structure of the music folder is kept intact
- **Added a warning message and log for unsupported music files detected and skipped**
    - Warns users that there are unsupported files which cannot be used anymore in their music folder (should there be any)
    - Logs the detected files to `music/unsupported_music_files.txt`
    
#### Minor Changes
- Remove the limit on searching subdirectories for the music folder
- Rename some functions to more in-line with what they're doing, reading files instead of looking for them
- Prettify and add more detail to the README, because it is old and could use some love
- Update prepackaged `.mmrs` files to use the new format

> [!IMPORTANT]
> There is no fallback for using either of the previously supported file formats. Maintaining support for them is antithetical to the intention and purpose of this update.

> [!CAUTION]
> Currently, the conversion script runs every time a seed is generated — the previous state of the music folder is backed up (it's replaced every time the script runs), and the folder will run through every file to check if it needs to be converted. If users have a large amount of music that needs to be converted — which is likely — the first conversion of files may take a while on the first run through the music folder. Subsequent runs should not take any longer than it usually does.
> 
> The conversion script is entirely self-contained within `MusicConversionUtils.cs` outside of the required call inside `builder.cs`. This allows it to easily be removed or disabled should it end up causing issues, or should an option be added to enable and disable it. (Modifying the GUI is something I don't want to mess with.)

### Testing
The following things were tested and were working properly in my debug build on my PC before making the pull request:
- Music conversion properly backs up the music folder before it replaces it
- Music conversion can handle files with multiple sequences and banks, and properly links them (it creates a separate archive for each entry)
- Music conversion properly mimics the structure of the original folder
- Music conversion properly copies, converts, and packs files from the old format to the new format
- Sequence files (`.seq`) are recognized and processed properly
- The new metadata file (`.meta`) is recognized and processed properly
    - Display names are properly read and set
    - Instrument sets are properly read and set
    - Categories are properly read and set
    - Commands are properly read and stored
- Bank files (`.zbank`) are recognized and processed properly
- Bankmeta files (`.bankmeta`) are recognized and processed properly
- Custom audio sample files (`.zsound`) are recognized and processed properly when the `ZSOUND` command is included in the `.meta` file
- Formmask files (`.formmask`) are recognized and processed properly
- `songtest` and `songforce` tokens continue to utilize the filename instead of the display name
- A warning message will pop up after seed generation if unsupported music files are detected
- Unsupported music files will be logged to a log file in the music folder if detected
- Music directory searches all subdirectories

### Impact of the Music Format Change
With the addition of a conversion utility into the randomizer itself, users will not have to do anything to ensure their music continues working with the randomizer — so the impact should be minimal. However, should it be needed a Python script that has nearly the same functionality can be found [in this repo](https://github.com/crinuleiroz/mmrs-format-updater) — sequencers are most likely to require this than users though (hopefully).

Commit:
Update MMRS format to use OOTRS meta files

This change updates the MMRS format to use OOTRS `.meta` files.

Along with this change, support for standalone sequences and the old MMRS format are removed entirely. Continuing their support only hampers further code updates.

`songtest` and `songforce` tokens will use the cosmetic name instead of the filename, but maybe they should retain old functionality...

Due to the way MMR handles categories, nothing about them has been changed currently.

Due to the way MMR handles ensuring music uses the correct sequence player, the song type line in the meta file is unused.

The meta file's format only requires 2 lines, in the case that there are less than 3 lines, defaults are applied for song type (bgm) and categories (0, 1, 3).

Because MMR doesn't parse the bank to link custom audio samples, the format for the ZSOUND command uses the old format.

Commit:
Add DisplayName field to SequenceInfo

Adds a new field to be used for the in-game display name

Commit:
Cosmetic name uses SequenceInfo's DisplayName instead of Name

Names displayed in-game will use the SequenceInfo's DisplayName field instead of using the filename without affecting plando, spoiler log, songforce and songtest, etc. that will continue using the filename.

Commit:
Use constant instead of explicit value

Commit:
Add all bgm categories to defaulted categories

Not really a point in limiting the default pool unless a category blatantly causes issues, which none of them theoretically should

Commit:
Add detection and logging for previously supported music files

Commit:
Add warning message for unsupported music files

Commit:
Create MusicConversionUtils.cs

Handles conversion of the music folder from the old format into the new format from within the randomizer itself. And after lots of testing... hopefully it works...

Commit:
Add old music to new music function call

Commit:
Prettify the README

Because the README is old and could use a fresh coat of paint.

Commit:
Update prepackaged music files

Update the music files that come prepackaged with the rando from the old format to the new format

Commit:
Update dead links

Commit:
Remove blurb bout zseq support

Commit:
Fix grammatical error in heading

ZoeyZolotova commented on MMR.Randomizer/Utils/SequenceUtils.cs:
```diff
@@ -417,73 +321,171 @@ private static void ScanForMMRSFormMask(SequenceInfo song, string sequenceName,
+                #endif
+            }
+
+            claimedBankCount += customBankIncluded;
```
I would like it if we changed `customBankIncluded` and the function it comes from to a bool, and change this line to
```
if (customBankIncluded) {
    claimedBankCount++;
}
```

ZoeyZolotova commented on MMR.Randomizer/Utils/SequenceUtils.cs:
```diff
@@ -417,73 +321,171 @@ private static void ScanForMMRSFormMask(SequenceInfo song, string sequenceName,
                 }
-                if (song.Instrument == MARK_REQUIRES_NEW_BANK && customBankIncluded == 0)
+            }
+            if (song.Instrument == MARK_REQUIRES_NEW_BANK && customBankIncluded = 0)
```
Should this be `==`? Anyway if we change this to a bool just do `&& !customBankIncluded)`

ZoeyZolotova commented on MMR.Randomizer/Utils/SequenceUtils.cs:
```diff
@@ -40,7 +40,8 @@ public class SequenceUtils
         public static MD5 md5lib; // used for zip
 
-
+        public static List<string> OLD_MUSIC_FILES = new List<string>(); // list to store the names of old music files found in the music directory
```
I don't see anywhere where you clear this array, so I think that means that once the warning pops up, it'll keep popping up each time you generate a seed until you close and reopen the program.

ZoeyZolotova commented on MMR.Randomizer/Utils/SequenceUtils.cs:
```diff
@@ -86,12 +87,19 @@ public static void ReadSequenceInfo()
             directories.AddRange(Directory.GetDirectories(Values.MusicDirectory).ToList());
             foreach (string d in directories.ToList()) // another layer deep to be safe
```
Can we skip this initial iteration of the folders in the music directory and just go straight for `Directory.GetDirectories(Values.MusicDirectory, "*", SearchOption.AllDirectories)`?

ZoeyZolotova commented on MMR.Randomizer/Builder.cs:
```diff
@@ -121,6 +121,8 @@ private void WriteAudioSeq(Random random, OutputSettings _settings)
                 return;
             }
 
+            MusicConversionUtils.ConvertMusicFiles(); // backs up music folder, copies music folder, converts any old files... might take a *long time* on the first pass if the user has a large folder
```
What will the behavior be if someone converts their music files via this process, then adds more "old format" files?

ZoeyZolotova reviewed (commented):
Thanks a lot for this! I've added a few comments. Also, this only makes the formats the same, right? It doesn't make them cross-compatible or anything does it? I was also wondering why we use a plaintext meta file; it's kind of hard to read in plaintext tbh. JSON or YML would be nicer, or really anything that describes what each line actually means.

@isghj5 would you mind reviewing this pull request too?

ghost commented:
Correct, this only makes the formats the files use the exact same. However, it is one of the steps required for getting cross-compatibility implemented.

I would also prefer JSON or YML, preferably YML as it's more human-readable, but the standard OOTR uses is the plaintext file — and this update is meant to ensure the files have the same standard/format. I could add a format like the following template:
```yml
---
game: oot # or mm

metadata:
  song name: "song name"
  instrument set: custom
  song type: bgm # or fanfare
  categories: [0x01, . . . ]
    
  audio samples:
    "majora.zsound": # mmr still doesn't parse the bank to link the sample and replaces based on address only
      instrument type: ~
      list index: ~
      key region: ~
      temp address: 0x20000427

# anything new or extra goes here
```

It would just require OOTR to update from their current format to whatever format would be implemented instead of the plaintext format. For now, the goal is just standardizing with the current OOTR format.

Commit:
Apply requested changes

Change "customBankIncluded" from int to a boolean, add a clear for OLD_MUSIC_FILES at the start of "ReadSequenceInfo", build iterable list of directories using IEnumerable.

Commit:
Cleanup ProcessFiles

Commit:
Use regex to remove songforce/songtest tokens from cosmetic names

Commit:
Fix regex
